### PR TITLE
Add the start of a Solid importer/exporter for contacts

### PR DESCRIPTION
**Pull request by bwillard**
See #531

### CODE REVIEW

File: extensions/data-transfer/portability-data-transfer-solid/src/main/java/org/datatransferproject/transfer/solid/SolidUtilities.java [hunk 1]
```diff
@@ -0,0 +1,229 @@
+package org.datatransferproject.transfer.solid;
```
Comment (holachuy) [inline extensions/data-transfer/portability-data-transfer-solid/src/main/java/org/datatransferproject/transfer/solid/SolidUtilities.java]: License header

File: extensions/data-transfer/portability-data-transfer-solid/src/main/java/org/datatransferproject/transfer/solid/SolidUtilities.java [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+import org.datatransferproject.transfer.solid.contacts.SolidContactsExport;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SolidUtilities {
+  private static final Logger logger = LoggerFactory.getLogger(SolidContactsExport.class);
+  private static final Pattern PROBLEMATIC_TURTLE = Pattern.compile("(\\s\\d+\\.)\n");
+  private static final HttpTransport TRANSPORT = new NetHttpTransport();
+  private final String authCookie;
+  final HttpRequestFactory factory;
```
Comment (holachuy) [inline extensions/data-transfer/portability-data-transfer-solid/src/main/java/org/datatransferproject/transfer/solid/SolidUtilities.java]: Is this member exposed publicly for testing or for other reasons. We should annotate for it's use.

File: extensions/data-transfer/portability-data-transfer-solid/src/main/java/org/datatransferproject/transfer/solid/SolidUtilities.java [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+    if (isType(selfResource, "http://www.w3.org/ns/ldp#Container")) {
+      for (Resource r : getContainedResource(model, url)) {
+        explore(r.getURI(), resourceConsumer);
+      }
+    }
+
+    resourceConsumer.accept(selfResource);
+  }
+
+  public Model getModel(String url) throws IOException {
```
Comment (holachuy) [inline extensions/data-transfer/portability-data-transfer-solid/src/main/java/org/datatransferproject/transfer/solid/SolidUtilities.java]: doc for public method here and below

File: extensions/data-transfer/portability-data-transfer-solid/src/main/java/org/datatransferproject/transfer/solid/SolidUtilities.java [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+    IOUtils.copy(response.getContent(), writer, "UTF-8");
+    String fixedString = fixProblematicPeriods(writer.toString());
+    Model defaultModel = ModelFactory.createDefaultModel();
+    return defaultModel.read(
+        new StringReader(fixedString),
+        url,
+        "TURTLE");
+  }
+
+  public static List<Statement> getProperties(Resource r, String property) {
```
Comment (holachuy) [inline extensions/data-transfer/portability-data-transfer-solid/src/main/java/org/datatransferproject/transfer/solid/SolidUtilities.java]: Maybe we can move statics to the bottom and make final.

File: extensions/data-transfer/portability-data-transfer-solid/src/main/java/org/datatransferproject/transfer/solid/SslHelper.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
+package org.datatransferproject.transfer.solid;
```
Comment (holachuy) [inline extensions/data-transfer/portability-data-transfer-solid/src/main/java/org/datatransferproject/transfer/solid/SslHelper.java]: license header

File: extensions/data-transfer/portability-data-transfer-solid/src/main/java/org/datatransferproject/transfer/solid/contacts/SolidContactsExport.java [hunk 6]
```diff
@@ -0,0 +1,237 @@
+package org.datatransferproject.transfer.solid.contacts;
```
Comment (holachuy) [inline extensions/data-transfer/portability-data-transfer-solid/src/main/java/org/datatransferproject/transfer/solid/contacts/SolidContactsExport.java]: license header

File: extensions/data-transfer/portability-data-transfer-solid/src/main/java/org/datatransferproject/transfer/solid/contacts/SolidContactsExport.java [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+    List<Statement> statements = SolidUtilities.getProperties(r, property);
+    checkState(statements.size() == 1,
+        "Expected just one item for resource %s, property: %s: %s",
+        r,
+        property,
+        statements);
+    return statements.get(0);
+  }
+
+  static VCard parsePerson(Resource r) {
```
Comment (holachuy) [inline extensions/data-transfer/portability-data-transfer-solid/src/main/java/org/datatransferproject/transfer/solid/contacts/SolidContactsExport.java]: final

File: extensions/data-transfer/portability-data-transfer-solid/src/main/java/org/datatransferproject/transfer/solid/contacts/SolidContactsImport.java [hunk 8]
```diff
@@ -0,0 +1,318 @@
+package org.datatransferproject.transfer.solid.contacts;
```
Comment (holachuy) [inline extensions/data-transfer/portability-data-transfer-solid/src/main/java/org/datatransferproject/transfer/solid/contacts/SolidContactsImport.java]: license header

File: extensions/data-transfer/portability-data-transfer-solid/src/main/java/org/datatransferproject/transfer/solid/contacts/SolidContactsImport.java [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SolidContactsImport implements Importer<CookiesAndUrlAuthData, ContactsModelWrapper> {
+  //See https://www.w3.org/TR/vcard-rdf/ for details.
+
+  private static final Logger logger = LoggerFactory.getLogger(SolidContactsExport.class);
+  private static final String TEST_SLUG_NAME = "ImportedAddressBook";
+  private static final String BASE_DIRECTORY = "/inbox/";
+  static final String IMPORTED_ADDRESS_BOOK_PATH = BASE_DIRECTORY + TEST_SLUG_NAME + "/";
```
Comment (holachuy) [inline extensions/data-transfer/portability-data-transfer-solid/src/main/java/org/datatransferproject/transfer/solid/contacts/SolidContactsImport.java]: move below the other private static members. Add annotation if it's package private for testing.

File: extensions/data-transfer/portability-data-transfer-solid/src/main/java/org/datatransferproject/transfer/solid/contacts/SolidContactsImport.java [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+      r.addProperty(VCARD4.hasPhoto, photo.getUrl());
+    }
+
+    StringWriter stringWriter = new StringWriter();
+    personModel.write(stringWriter, "TURTLE");
+    return stringWriter.toString();
+  }
+
+  /** Looks up the {@link Resource}s for a given string, that might be comma delimited. **/
+  private static ImmutableList<Resource> getPhoneOrMailTypes(
```
Comment (holachuy) [inline extensions/data-transfer/portability-data-transfer-solid/src/main/java/org/datatransferproject/transfer/solid/contacts/SolidContactsImport.java]: final

File: extensions/data-transfer/portability-data-transfer-solid/src/test/java/org/datatransferproject/transfer/solid/contacts/ManualTest.java [hunk 11]
```diff
@@ -0,0 +1,53 @@
+package org.datatransferproject.transfer.solid.contacts;
```
Comment (holachuy) [inline extensions/data-transfer/portability-data-transfer-solid/src/test/java/org/datatransferproject/transfer/solid/contacts/ManualTest.java]: license header

File: extensions/data-transfer/portability-data-transfer-solid/src/test/java/org/datatransferproject/transfer/solid/contacts/RoundTripToRdfTest.java [hunk 12]
```diff
@@ -0,0 +1,74 @@
+package org.datatransferproject.transfer.solid.contacts;
```
Comment (holachuy) [inline extensions/data-transfer/portability-data-transfer-solid/src/test/java/org/datatransferproject/transfer/solid/contacts/RoundTripToRdfTest.java]: license header

File: extensions/data-transfer/portability-data-transfer-solid/src/test/java/org/datatransferproject/transfer/solid/contacts/TestData.java [hunk 13]
```diff
@@ -0,0 +1,94 @@
+package org.datatransferproject.transfer.solid.contacts;
```
Comment (holachuy) [inline extensions/data-transfer/portability-data-transfer-solid/src/test/java/org/datatransferproject/transfer/solid/contacts/TestData.java]: license header

File: portability-types-transfer/src/main/java/org/datatransferproject/types/transfer/auth/CookiesAndUrlAuthData.java [hunk 14]
```diff
@@ -0,0 +1,29 @@
+package org.datatransferproject.types.transfer.auth;
```
Comment (holachuy) [inline portability-types-transfer/src/main/java/org/datatransferproject/types/transfer/auth/CookiesAndUrlAuthData.java]: license header

File: extensions/data-transfer/portability-data-transfer-solid/src/main/java/org/datatransferproject/transfer/solid/contacts/SolidContactsImport.java (hunk 10 above)
Comment (bwillard) [reply extensions/data-transfer/portability-data-transfer-solid/src/main/java/org/datatransferproject/transfer/solid/contacts/SolidContactsImport.java]: Why mark a private static method final, it has to be final.  (I am agnostic, but IntelliJ complains if you mark them final, which makes me think doing so might be an anti-pattern)

**Comment by bwillard** [comment]
Hey Chuy, thanks for the comments, PTAL

**Pull request merged**